File: exercism/elixir/bob/lib/bob.ex
defmodule Bob do
  @spec hey(String.t()) :: String.t()
  def hey(input) do
    input = input |> String.trim()
    yell = input == String.upcase(input) and input != String.downcase(input)
    question = input |> String.ends_with?("?")

    cond do
      input == "" -> "Fine. Be that way!"
      yell and question -> "Calm down, I know what I'm doing!"
      yell -> "Whoa, chill out!"
      question -> "Sure."
      true -> "Whatever."
    end
  end
end

File: exercism/cpp/bob/bob.cpp
#include "bob.h"
#include <regex>

namespace bob {

std::string hey(const std::string speech) {
    bool uppercase = false, 
         lowercase = false,
         question  = false,
         blank     = true;

    for (auto c: speech) {
        if (isupper(c)) uppercase = true;
        if (islower(c)) lowercase = true;
        if (!isspace(c)) blank = question = false;
        if (c == '?') question = true;
    }

    bool yelling = uppercase && !lowercase;

    if (blank) return "Fine. Be that way!";
    if (yelling && question) return "Calm down, I know what I'm doing!";
    if (question) return "Sure.";
    if (yelling) return "Whoa, chill out!";
    return "Whatever.";
}

}
